File: src/main/java/com/hp/jipp/model/Code.java
package com.hp.jipp.model;

import com.google.auto.value.AutoValue;
import com.hp.jipp.encoding.NameCode;
import com.hp.jipp.encoding.NameCodeType;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/** A superset of Status and Operation */
@AutoValue
public abstract class Code extends NameCode {

    private static final List<Code> ALL;

    static {
        List<Code> codes = new ArrayList<>();

        for (NameCode status : Status.ENCODER.getMap().values()) {
            codes.add(new AutoValue_Code(status.getName(), status.getCode()));
        }
        for (NameCode operation : Operation.ENCODER.getMap().values()) {
            codes.add(new AutoValue_Code(operation.getName(), operation.getCode()));
        }
        ALL = Collections.unmodifiableList(codes);
    }

    public static final NameCodeType.Encoder<Code> ENCODER = NameCodeType.Encoder.of(
            "Code", ALL, new NameCode.Factory<Code>() {
                @Override
                public Code of(String name, int code) {
                    return new AutoValue_Code(name, code);
                }
            });
}

File: src/main/java/com/hp/jipp/model/Code.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.model

import com.hp.jipp.encoding.Enum
import com.hp.jipp.encoding.EnumType

/** A superset of Status and Operation  */
abstract class Code : Enum() {

    companion object {
        // MUST be lazy because Status and Operation are subclasses.
        val Encoder by lazy {
            EnumType.Encoder("Code", Status.Encoder.map.values + Operation.Encoder.map.values) {
                code: Int, name: String -> Status(code, name)
            }
        }
    }
}
